### PR TITLE
Fix decomposition of controlled quantum gates

### PR DESCRIPTION
**Pull request by nikhilkhatri**

### CODE REVIEW

File: discopy/quantum/gates.py [hunk 1]
```diff
@@ -365,13 +365,27 @@ def _decompose_grad(self):
...
             return self
-        src, tgt = (0, 1) if distance > 0 else (1, 0)
-        perm = Circuit.permutation([src, *range(2, n_qubits), tgt])
-        diagram = (perm[::-1]
-                   >> type(self)(controlled) @ Id(abs(distance) - 1)
-                   >> perm)
+
+        skipped_qbs = n_qubits - (1 + len(controlled.dom))
+
+        if (distance > 0):
```
Comment (ianyfan) [inline discopy/quantum/gates.py:374]: Remove brackets

**Review by y-richie-y — APPROVED** [review]
LGTM

File: discopy/quantum/gates.py [hunk 2]
```diff
@@ -365,13 +365,27 @@ def _decompose_grad(self):
...
             return self
-        src, tgt = (0, 1) if distance > 0 else (1, 0)
-        perm = Circuit.permutation([src, *range(2, n_qubits), tgt])
+
+        skipped_qbs = n_qubits - (1 + len(controlled.dom))
+
+        if distance > 0:
+            pattern = [0,
+                       *range(skipped_qbs + 1,
+                              len(controlled.dom) + skipped_qbs + 1),
```
Comment (ianyfan) [inline discopy/quantum/gates.py:377]: Can't this be  `n_qubits`?

File: discopy/quantum/gates.py [hunk 3]
```diff
@@ -365,13 +365,27 @@ def _decompose_grad(self):
...
+                       *range(1, skipped_qbs + 1)]
+        else:
+            pattern = [n_qubits - 1,
+                       *range(len(controlled.dom)),
+                       *range(len(controlled.dom), n_qubits - 1)]
```
Comment (ianyfan) [inline discopy/quantum/gates.py:382]: Can't you combine these ranges?

**Review by ianyfan — COMMENTED** [review]
Some comments about the ranges, otherwise looks good

**Pull request merged**